core/src/main/java/org/infinispan/interceptors/distribution/TriangleDistributionInterceptor.java

```diff
@@ -232,24 +237,33 @@ private BasicInvocationStage primaryOwnerWrite(InvocationContext context, DataWr
             }
             return;
          }
-         if (distributionInfo.owners().size() > 1) {
-            Collection<Address> backupOwners = distributionInfo.backups();
-            if (rCtx.isOriginLocal() && (isSynchronous(dwCommand) || dwCommand.isReturnValueExpected())) {
-               commandAckCollector.create(id, rv, distributionInfo.owners(), dwCommand.getTopologyId());
-            }
-            if (trace) {
-               log.tracef("Command %s send to backup owner %s.", dwCommand.getCommandInvocationId(), backupOwners);
-            }
+         //protect with the shared lock.
+         //it was possible the topology change before the collector was created. the collector would be initializer
+         //with an old members that would never send the ack back resulting in timeout.
+         stateTransferLock.acquireSharedTopologyLock();
```

Comment:
I would prefer checking the topology again after the collector was created, and retry if it changed, to avoid the contention on the number of readers in the shared lock.